RFDs/0001-job-worker-service.md

```diff
@@ -0,0 +1,468 @@
+---
+authors: Zeeshan Haque (@zphoenixrises)
+state: review
+---
+
+# RFD 0001 - Job Worker Service
+
+## What
+
+This RFD proposes a prototype job worker service that provides a secure API to run arbitrary Linux processes with fine-grained resource control and real-time output streaming. The service aims to demonstrate a scalable and secure approach to process execution in a multi-tenant environment.
+
+## Why
+
+To showcase a secure and efficient approach to running and managing arbitrary processes in a controlled environment.
+
+## Design Overview
+
+### Components
+
+1. **Library**: A reusable job management library
+2. **API**: A gRPC API server
+3. **Client**: A command-line interface (CLI) client
+
+### Key Features
+
+- Job management (start, stop, query status, get output)
+- Real-time output streaming with support for multiple concurrent clients
+- Resource control using cgroups (CPU, Memory, Disk I/O)
+- Secure communication using mTLS
+- Simple role-based authorization scheme
+- gRPC API for client-server communication
+
+#### Key Considerations 
+- This design is not a distributed system and is only optimized to run on a single linux OS
+- If needed, the components of the library can be split into a distributed microservices framework inthe future
+- Rudimentary monitoring of total resource usage will be available through the the CLI if needed.
+
+## Detailed Design
+
+### 1. Library
+
+#### a. In-memory Job Logger
+- Implement a thread-safe circular buffer to store logs with a configurable maximum size
+- Use a read-write mutex to ensure thread-safety for concurrent read/write operations
+- Implement a pub/sub pattern for real-time log streaming:
+  - Use channels for each subscriber
+  - Implement a fan-out pattern to broadcast new log entries to all subscribers
+- Provide methods for adding logs, retrieving all logs, and subscribing to log streams
+- Implement log rotation to handle long-running jobs and prevent memory exhaustion
+
+```go
+// Logger represents an in-memory circular buffer for storing job logs.
+type Logger struct {
+	mu          sync.RWMutex
+	logs        []string
```

Comment:
Is `string` the best data type of arbitrary command output?

Reply:
You are right. I had not considered binary output. I changed it to bytes.Buffer to be able to handle any kind of output of arbitrary length.